src/ray/common/scheduling/cluster_resource_scheduler.h

```diff
@@ -343,25 +356,46 @@ class ClusterResourceScheduler {
   ///
   /// \param available A list of available capacities for resource's instances.
   /// \param resource_instances List of the resource instances being updated.
-  void AddAvailableResourceInstances(std::vector<double> available,
-                                     ResourceInstanceCapacities *resource_instances);
+  ///
+  /// \return Overflow capacities of "resource_instances" after adding instance 
+  /// capacities in "available". 
+  std::vector<double> AddAvailableResourceInstances(std::vector<double> available,
+                                                    ResourceInstanceCapacities *resource_instances);
 
   /// Decrease the available capacities of the instances of a given resource.
   ///
   /// \param free A list of capacities for resource's instances to be freed.
   /// \param resource_instances List of the resource instances being updated.
-  void SubtractAvailableResourceInstances(std::vector<double> free,
-                                          ResourceInstanceCapacities *resource_instances);
+  /// \return Under capacities of "resource_instances" after substracting instance 
+  /// capacities in "free". 
+  std::vector<double> SubtractAvailableResourceInstances(std::vector<double> free,
+                                                         ResourceInstanceCapacities *resource_instances);
 
   /// Increase the available CPU instances of this node.
   ///
   /// \param cpu_instances CPU instances to be added to available cpus.
-  void AddCPUResourceInstances(std::vector<double> &cpu_instances);
+  ///
+  /// \return Overflow capacities of CPU instances after adding CPU 
+  /// capacities in cpu_instances. 
+  std::vector<double> AddCPUResourceInstances(std::vector<double> &cpu_instances);
 
-  /// Decrease the available cpu instances of this node.
+  /// Decrease the available CPU instances of this node.
   ///
-  /// \param cpu_instances Cpu instances to be removed from available cpus.
-  void SubtractCPUResourceInstances(std::vector<double> &cpu_instances);
+  /// \param cpu_instances CPU instances to be removed from available cpus.
+  ///
+  /// \return Underflow capacities of CPU instances after subtracting CPU 
+  /// capacities in cpu_instances. 
+  std::vector<double> SubtractCPUResourceInstances(std::vector<double> &cpu_instances);
+
+  bool AllocateTaskResources(int64_t node_id, const TaskRequest &task_req, 
```

Comment:
Document. Same for below methods.

Also, do we really need 3 different methods for allocating task resources? Maybe one of these should be a private method?

Reply:
Good point. AllocatedTaskResources() is private now.